base/linalg/bidiag.jl

```diff
@@ -108,7 +108,7 @@ ctranspose(M::Bidiagonal) = Bidiagonal(conj(M.dv), conj(M.ev), !M.isupper)
 istriu(M::Bidiagonal) = M.isupper || all(M.ev .== 0)
 istril(M::Bidiagonal) = !M.isupper || all(M.ev .== 0)
 
-function tril(M::Bidiagonal, k::Integer=0)
+function tril!(M::Bidiagonal, k::Integer=0)
```

Comment:
I think this applies throughout for the rest of the types too. Anywhere it's possible to make the `!` versions allocation-free and in-place, then I think that would be a good goal. And aim for type-stability but with the smallest amount of type widening that makes sense. The one-arg case could possibly be made to return a more specialized type than the two-arg case for some types since it's more specific in its behavior, but that would require extra methods rather than using default argument values and could be left as a future enhancement.